documentation/sphinx/source/internal-dev-tools.rst

```diff
@@ -0,0 +1,58 @@
+##################
+Internal Dev Tools
+##################
+
+Code Probes
+===========
+
+Code probes are a mechanism in FDB to prove that certain code-paths are being tested under the right conditions. They differ from code coverage in multiple ways (explained below).
+
+The general format of a code probe is:
+
+.. code-block:: C++
+
+   CODE_PROBE(<condition>, "Comment", [annotations...]);
+
+A simple example of a code probe could look as follows:
+
+.. code-block:: C++
+
+   CODE_PROBE(self->forceRecovery, "Resolver detects forced recovery", context::sim2);
+
+On a very high level, the above code will indicate that whenever this line is executed and ``self->forceRecovery`` is ``true``, we ran into some interesting case. In addition this probe is also annotated with ``context::sim2``. This indicates that we expect this code to be eventually hit in simulation.
+
+By default, FDB simply will write a trace-line when this code is hit and the condition is ``true``. If the code is never hit, the simulator will, at the end of the run, print the code probe but set the ``covered`` field to ``false``.
+
+We expect that ALL code probes will be hit in a nightly run. In the future we can potentially use this feature for other things (like instructing the simulator to do an extensive search starting when one of these probes is being hit).
+
+In addition to ``context`` annotations, users can also define and pass assertsions. For example:
+
+.. code-block:: C++
+
+   CODE_PROBE(condition, "Some comment", assert::simOnly);
+
+These will add an assertion to the code. In addition to that, the simulator will not print missed code probes that asserted that the probe won't be hit in simulation.
+
+Test Harness
+============
+
+TestHarness is our primary testing tool. It has multiple jobs:
+
+* *Running*: It can run a test in Joshua.
+* *Statictics*: It will chose a test to run based on previously spent CPU time for all tests. It does that by writing statistics about the test at the end of each run.
```

Comment:
It would be useful to say what 'previously' spans, i.e., previous runs of a test in an ensemble, versus history of ensemble runs in a joshua installation. Also, where are the statistics written? `based on` is vague. As a test writer, I don't know how I should think about my test design given this information.